invoke-ai/InvokeAI
# feat(ui): toasts rework

## Summary

Add a utility for managing toasts. A single method `toast` is exported. It can be used to both create a new toast or update an existing one, if provided the ID of an existing toast. All chakra toast options are supported, plus one new option: `withCount`.

`withCount` defaults to true, and handles appending a couple to the title of the toast when mutltiples are fired:
```ts
toast({ title: 'Hello', withCount: true });
// first toast title: 'Hello'
// second toast title: 'Hello (2)'
// third toast title: 'Hello (3)'
``` 

When updating a toast, the new title, description and other params will immediately override the existing toast.

If set to false, no count is appended. Note that if the id changes each time the toast is fired, that is considered a different toast and it will create a new one instead of updating.

All toasts across the app have been revised to use this utility. There's no more "toast queue" - toasts are fired the same way whether in a component, listener or wherever.

Other changes:
- Fixed recalling seed not showing a toast
- Consolidated some model install logic that used toasts, was duplicated like several times

## Related Issues / Discussions

- Supersedes #6409 - I copied the component and handling of invocation error events from this PR, thanks @maryhipp 
- Closes #5136
- Closes #4637

## QA Instructions

Note: There is some minor jank with how the toasts resize if you update them rapidly. This is a chakra issue - will need to dig into chakra theming to fix it.

- Do something that makes a toast many times in a row. For example, spam one of the parameter recall buttons.
- Create a graph with a single divide integer node. Leave the values at zero to create a divide-by-zero error. Increase iterations to 5 or so and invoke. You should see one batch queued toast and one error toast that rapidly updates.

## Merge Plan

n/a

## Checklist

- [ ] _Tests added / updated (if applicable)_ This needs proper e2e tests to test, but I don't think this particular functionality is super important to test.
